### DISCUSSION TITLE
I know this isn't SO, but it's a failry niche space and someone might be able to answer this

### DISCUSSION THREAD

**Discussion by rlewkowicz** [x360ce/x360ce#1499] [Q&A] [answered]
TLDR: Is the 360 controller right joystick x/y 16 bits?

I dumped the report descriptor for a 360 controller and I'm still trying to understand what I'm looking at. 

```
0x05, 0x01,        // Usage Page (Generic Desktop Ctrls)
0x09, 0x05,        // Usage (Game Pad)
0xA1, 0x01,        // Collection (Application)
0x09, 0x00,        //   Usage (Undefined)
0xA1, 0x00,        //   Collection (Physical)
0x09, 0x30,        //     Usage (X)
0x15, 0x00,        //     Logical Minimum (0)
0x35, 0x00,        //     Physical Minimum (0)
0x65, 0x00,        //     Unit (None)
0x55, 0x00,        //     Unit Exponent (0)
0x75, 0x10,        //     Report Size (16)
0x95, 0x01,        //     Report Count (1)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x31,        //     Usage (Y)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC1, 0x00,        //   End Collection
0x09, 0x00,        //   Usage (Undefined)
0xA1, 0x00,        //   Collection (Physical)
0x09, 0x33,        //     Usage (Rx)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x34,        //     Usage (Ry)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC1, 0x00,        //   End Collection
0x09, 0x00,        //   Usage (Undefined)
0xA1, 0x00,        //   Collection (Physical)
0x09, 0x32,        //     Usage (Z)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC1, 0x00,        //   End Collection
0x05, 0x09,        //   Usage Page (Button)
0x19, 0x01,        //   Usage Minimum (0x01)
0x29, 0x0A,        //   Usage Maximum (0x0A)
0x25, 0x01,        //   Logical Maximum (1)
0x45, 0x01,        //   Physical Maximum (1)
0x75, 0x01,        //   Report Size (1)
0x95, 0x0A,        //   Report Count (10)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x05, 0x01,        //   Usage Page (Generic Desktop Ctrls)
0x09, 0x39,        //   Usage (Hat switch)
0x15, 0x01,        //   Logical Minimum (1)
0x25, 0x08,        //   Logical Maximum (8)
0x46, 0x3B, 0x10,  //   Physical Maximum (4155)
0x65, 0x0E,        //   Unit (None)
0x75, 0x04,        //   Report Size (4)
0x95, 0x01,        //   Report Count (1)
0x81, 0x42,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,Null State)
0x15, 0x00,        //   Logical Minimum (0)
0x25, 0x01,        //   Logical Maximum (1)
0x45, 0x01,        //   Physical Maximum (1)
0x65, 0x00,        //   Unit (None)
0x75, 0x01,        //   Report Size (1)
0x95, 0x12,        //   Report Count (18)
0x81, 0x03,        //   Input (Const,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC1, 0x00,        // End Collection

// 107 bytes

// best guess: USB HID Report Descriptor
```

Right here is what I'm trying to figure out:
```
0x09, 0x33,        //     Usage (Rx)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x34,        //     Usage (Ry)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
```
Lets move the right joystick all the way up and actually look at the report that is sent:
```
00 14 00 00 00 00 00 00 00 00 a1 3f ff 7f 00 00 00 00 00 00
```
`a1 3f ff 7f` is our right joystick. Now I can decode a portion of this into 8 bits and get the expected -127/127 etc. But there's 2 other bits that move. So you'd think it's 16, but I think 30k is supposed to be the middle and in my testing it just doesn't line up right. 

Can anyone tell me what `a1 3f ff 7f` is?

**Comment by EvoJocys** [comment, marked as answer]
Yes, 16 bits.

Report Size (16) and Report Count (1) are global items. They are set once before Usage (X) and never changed afterwards, so X, Y, Rx, Ry and Z all inherit them. That is why no size appears next to Rx and Ry.

The report you captured is the 20 byte XUSB layout rather than a plain HID one:

```
byte 0       message type (00)
byte 1       length (0x14 = 20)
bytes 2-3    buttons
byte 4       left trigger
byte 5       right trigger
bytes 6-7    left stick X
bytes 8-9    left stick Y
bytes 10-11  right stick X
bytes 12-13  right stick Y
```

So `a1 3f ff 7f` is bytes 10 to 13, which is Rx and Ry, each signed 16 bit little endian:

```
a1 3f  ->  0x3FA1  ->  16289
ff 7f  ->  0x7FFF  ->  32767
```

The centre is what is tripping you up. It is 0, not around 32768. The range is -32768 to 32767, so 0x7FFF is full deflection one way and 0x8000 is full deflection the other. Your -127 to 127 reading came from looking at the high byte on its own, which is the same value scaled down to 8 bits.

x360ce uses the same convention internally: thumb axes are signed 16 bit, triggers are 0 to 255.

Kind Regards,
Evo Jocys
